### DISCUSSION TITLE
cephfs quota set

### DISCUSSION THREAD

**Discussion by QZB349549482** [ceph/go-ceph#826] [Q&A] [answered]
In go-ceph branch master, I have not see a function to set quota limit for director . Please give me a example to 
solve this. 
PS: I want to make a dir and set quota limit on it by cephfs

**Comment by phlogistonjohn** [comment]
It's not clear to me what you are asking for.

I don't know what director is. The "PS" indicates you want to set a quota for cephfs. This is more helpful. What would be most helpful is if you could link to documentation explaining what you are trying to do. So for example if there's a ceph command line tool that does what you want, link to docs showing the use of the tool and then we can backtrack from the tool to what APIs may be needed.

Thank you.

**Reply by QZB349549482** [reply]
Thank you for your reply！ I get some information from ceph doc. 
```
setfattr -n ceph.quota.max_bytes -v 100000000 /some/dir     # 100 MB
setfattr -n ceph.quota.max_files -v 10000 /some/dir         # 10,000 files
```
i try to set the quota limit like this ,and use the Setfattr function from  the cephfs package ,but the param is not easy use ,and  it is not work . And I have tried to use MdsCommand to setQuota ,but the example in the command_test.go is not easy to understand.

**Reply by phlogistonjohn** [reply]
> Thank you for your reply！ I get some information from ceph doc.
> 
> ```
> setfattr -n ceph.quota.max_bytes -v 100000000 /some/dir     # 100 MB
> setfattr -n ceph.quota.max_files -v 10000 /some/dir         # 10,000 files
> ```

That's very helpful thanks.

> 
> i try to set the quota limit like this ,and use the Setfattr function from the cephfs package ,but the param is not easy use ,and it is not work 

This function?
https://github.com/ceph/go-ceph/blob/312e4cf6b5cfb02e69768232d0fb0888a7ba51ba/cephfs/path_xattr.go#L19-L27

Could you please elaborate on why you find it difficult to use? What errors did you get if you saw errors? 


> And I have tried to use MdsCommand to setQuota ,but the example in the command_test.go is not easy to understand.

Constructing the correct JSON object to use with an MdsCommand function can be tricky. However, since the example above uses `setfattr`, I don't know what `ceph` command this should be equivalent to.  Besides, since we have a SetXattr cephfs function I would think it makes more sense to use that.

**Comment by QZB349549482** [comment]
My most pressing need is this ，i want to set a quota limit on a directory. the flow is that  I create dir by mount.MakeDir which is in /go-ceph/cephfs/path.go, then i search the ceph doc for info :
```
setfattr -n ceph.quota.max_bytes -v 100000000 /some/dir     # 100 MB
setfattr -n ceph.quota.max_files -v 10000 /some/dir         # 10,000 files
```
I try to use the function in /go-ceph/cephfs/file_xattr.go
```
func (f *File) SetXattr(name string, value []byte, flags XattrFlags) error {
	if err := f.validate(); err != nil {
		return err
	}
	if name == "" {
		return errInvalid
	}
	var vptr unsafe.Pointer
	if len(value) > 0 {
		vptr = unsafe.Pointer(&value[0])
	}
	cName := C.CString(name)
	defer C.free(unsafe.Pointer(cName))

	ret := C.ceph_fsetxattr(
		f.mount.mount,
		f.fd,
		cName,
		vptr,
		C.size_t(len(value)),
		C.int(flags))
	return getError(ret)
}
```
and function in /cephfs/path_xattr.go
```
func (mount *MountInfo) LsetXattr(path, name string, value []byte, flags XattrFlags) error {
	if err := mount.validate(); err != nil {
		return err
	}
	if name == "" {
		return errInvalid
	}
	var vptr unsafe.Pointer
	if len(value) > 0 {
		vptr = unsafe.Pointer(&value[0])
	}
	cPath := C.CString(path)
	defer C.free(unsafe.Pointer(cPath))
	cName := C.CString(name)
	defer C.free(unsafe.Pointer(cName))

	ret := C.ceph_lsetxattr(
		mount.mount,
		cPath,
		cName,
		vptr,
		C.size_t(len(value)),
		C.int(flags))
	return getError(ret)
}
```
The problem is that I do not know the function is right to do it （PS: set a quota limit which get from https://docs.ceph.com/en/quincy/cephfs/quota/?highlight=setfattr%20-n%20ceph.quota.max_bytes%20-v%20100000000#configuration）,if it is right ,how can I construct a param to fill It .

**Comment by phlogistonjohn** [comment, marked as answer]
So, the setfattr tool will end up calling linux syscalls like setxattr & lsetxattr. As you've determined the cephfs api calls are similarly named. The tool's `-n` option corresponds to the name argument. The tool's `-v` argument corresponds to the value argument.  Since no special arguments are being passed for the value or encoding, you simply want to convert the value (
100000000, 10000) directly as a string, and convert that to a byte slice.  Roughly, that'd be:
```go
path := "/your/path/here"
name := "ceph.quota.max_bytes"
value := []byte("10000")
err := mount.LsetXattr(path, name, value, XattrDefault)
// ...handle errors...
```

You can also pass flags if you need them, but generally the default (0) is fine.
https://github.com/ceph/go-ceph/blob/312e4cf6b5cfb02e69768232d0fb0888a7ba51ba/cephfs/file_xattr.go#L24-L33

If this approach doesn't work please let us know and share any errors that you get.

**Reply by QZB349549482** [reply]
It is work ,thank you !
